### PR TITLE
feat: add EVM support

### PR DESCRIPTION
**Pull request by 0xApotheosis**
## Description

- Requires lib PR: https://github.com/shapeshift/lib/pull/902 before CI will pass.
- Includes cherry-picked commits from https://github.com/shapeshift/web/pull/2208.

Adds EVM support to web, supporting the following functionality:

- [x] Show Avalanche assets in the portfolio
- [x] Show Avalanche transactions
- [x] Use swappers supported by the Avalanche chain
- [x] Use AVAX as the native fee asset
- [x] Enable send/receive for EVM assets
- [x] Send fee from unchained too low to send - separate ticket: https://github.com/shapeshift/web/issues/2200
- [ ] Reduce account fetching page size 100 -> 10 until https://github.com/shapeshift/unchained/issues/460 is resolved

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Closes https://github.com/shapeshift/web/issues/2156

## Risk

Medium - touches chain/account specifier logic.

## Testing

Test both Ethereum and Avalanche EVM chain usage works as expected throughout the app.

## Screenshots (if applicable)

N/A

### CODE REVIEW

**Comment by gomesalexandre** [comment]
`Gracefully handle switching between EVM chains (Ethereum and Avalanche for now)` is already implemented in https://github.com/shapeshift/lib/pull/883 (automagic chain switch depending on the Tx that's being built) and will be getting some nice-to-have user improvements with https://github.com/shapeshift/web/pull/2165 to allow users to see/switch their wallet's EVM chain

File: src/lib/web3-instance.ts [hunk 1]
```diff
@@ -1,14 +1,15 @@
+import { ChainId } from '@shapeshiftoss/caip'
 import Web3 from 'web3'
 import { getWeb3Provider } from 'lib/web3-provider'
 
-let maybeWeb3Instance: Web3 | null
+const web3InstanceMap: Map<ChainId, Web3> = new Map()
```
Comment (0xApotheosis) [inline src/lib/web3-instance.ts:5]: We now use a map of `web3` instances, as we need one for each RPC URL, which is approximately one per EVM chain.

**Review by 0xApotheosis — COMMENTED** [review]
Notes for reviewers.

File: src/hooks/useEvm/useEvm.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { CHAIN_NAMESPACE, fromChainId } from '@shapeshiftoss/caip'
+import { useMemo } from 'react'
+import { getChainAdapters } from 'context/PluginProvider/PluginProvider'
+import { selectFeatureFlags } from 'state/slices/preferencesSlice/selectors'
+import { useAppSelector } from 'state/store'
+
+export const useEvm = () => {
```
Comment (gomesalexandre) [inline src/hooks/useEvm/useEvm.ts:7]: Is this ever going to do something else than return `supportedEvmChainIds` ?

File: src/lib/web3-provider.ts [hunk 3]
```diff
@@ -1,13 +1,26 @@
...
-let maybeWeb3Provider: InstanceType<typeof Web3.providers.HttpProvider> | null
+type HttpProvider = InstanceType<typeof Web3.providers.HttpProvider>
 
-export const getWeb3Provider = () => {
-  if (!maybeWeb3Provider) {
-    maybeWeb3Provider = new Web3.providers.HttpProvider(getConfig().REACT_APP_ETHEREUM_NODE_URL)
-    return maybeWeb3Provider!
+const web3ProviderMap: Map<ChainId, HttpProvider> = new Map()
+
+const httpProviderByChain = (chainId: ChainId): string => {
```
Comment (gomesalexandre) [inline src/lib/web3-provider.ts:9]: If we want to honor the parameter name
```suggestion
const httpProviderByChainId = (chainId: ChainId): string => {
```

File: src/lib/web3-provider.ts [hunk 4]
```diff
@@ -1,13 +1,26 @@
...
+  switch (chainId) {
+    case avalancheChainId:
+      return getConfig().REACT_APP_AVALANCHE_NODE_URL
+    case ethChainId:
+    default:
+      return getConfig().REACT_APP_ETHEREUM_NODE_URL
+  }
+}
+
+export const getWeb3Provider = (chainId: ChainId): HttpProvider => {
```
Comment (gomesalexandre) [inline src/lib/web3-provider.ts:19]: That might be better suited now that this is programmatic
```suggestion
export const getWeb3ProviderByChainId = (chainId: ChainId): HttpProvider => {
```

**Review by gomesalexandre — COMMENTED** [review]
- Tested for regressions in ETH sends / swaps and found none
- Also tested for valid pairs with the Avalanche flags and related lib PR linked, which seems to handle pairs from the same ChainId, as it
  - properly narrows down the available assets to swaps to Ethereum ones when on Ethereum / Avalanche ones when on Avalanche
  - pops a toast with invalid pair message when currently swapping from AVAX / Avalanche WETH and then switching the sell asset to ETH, which invalidates the pair
  
I requested C-Chain testing fund for further testing, will test Avalanche sends/swaps then

**Review by gomesalexandre — APPROVED** [review]
Functionally this seems to work as intended for Avax - tested swaps and sends and didn't see any weird lil bug.
Also regression tested Ethereum features (swaps/sends/FOX and Yearn vaults staking features) and didn't find any regression there either, both with the flag on and off.

Given that most of my comments are nitpicky, I'm going to early stamp this one - will re-stamp it when it gets invalidated by package bumps and possibly other new commits.

File: src/hooks/useEvm/useEvm.ts (hunk 2 above)
Comment (0xApotheosis) [reply src/hooks/useEvm/useEvm.ts:7]: I suspected that it would, but I don't know what yet.

File: src/lib/web3-provider.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/lib/web3-provider.ts]: Addressed in https://github.com/shapeshift/web/pull/2164/commits/aa5aaf46b2f8dc0b625bff71fb1c1c6008b93cf1

File: src/lib/web3-provider.ts (hunk 4 above)
Comment (0xApotheosis) [reply src/lib/web3-provider.ts]: Addressed in https://github.com/shapeshift/web/pull/2164/commits/aa5aaf46b2f8dc0b625bff71fb1c1c6008b93cf1

**Review by gomesalexandre — APPROVED** [review]
Tested again with the bumped packages installed and the previous scenario, still working as intended

🍦

File: src/lib/web3-instance.ts [hunk 5]
```diff
@@ -1,14 +1,15 @@
...
 
-export const getWeb3Instance = (): Web3 => {
-  const web3Provider = getWeb3Provider()
-  if (!maybeWeb3Instance) {
-    maybeWeb3Instance = new Web3(web3Provider)
-    return maybeWeb3Instance!
+export const getWeb3Instance = (chainId: ChainId): Web3 => {
```
Comment (kaladinlight) [inline src/lib/web3-instance.ts:7]: `getWeb3InstanceByChainId`

File: src/state/slices/assetsSlice/selectors.ts [hunk 6]
```diff
@@ -72,7 +72,9 @@ const chainIdFeeAssetReferenceMap = (
             return ASSET_REFERENCE.Bitcoin
         }
       case CHAIN_NAMESPACE.Ethereum:
-        return ASSET_REFERENCE.Ethereum
+        return chainReference === CHAIN_REFERENCE.AvalancheCChain
```
Comment (kaladinlight) [inline src/state/slices/assetsSlice/selectors.ts:75]: i have been updating these to a switch with exhaustive cases and a default throw for an invalid chain reference to be more explicit than default fallback

**Review by kaladinlight — DISMISSED** [review]
couple small comments, but looks solid overall!

File: src/lib/web3-instance.ts (hunk 5 above)
Comment (0xApotheosis) [reply src/lib/web3-instance.ts]: Good spot, addressed in https://github.com/shapeshift/web/pull/2164/commits/57f023bf9cb13475460440a417d188e49824bae4

File: src/state/slices/assetsSlice/selectors.ts (hunk 6 above)
Comment (0xApotheosis) [reply src/state/slices/assetsSlice/selectors.ts]: Yeah that is more extensible. for sure, updated in https://github.com/shapeshift/web/pull/2164/commits/17072ffbb088ec0687b9969411df34e7b08c98c9

**Review by 0xdef1cafe — APPROVED** [review]

**Review by gomesalexandre — APPROVED** [review]
Untested but approving as previously tested and approved and this is just bringing syntactical changes.

Thanks for the exhaustiveness and error-handling for all branches in https://github.com/shapeshift/web/pull/2164/commits/17072ffbb088ec0687b9969411df34e7b08c98c9 ser! 🥇

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**